### ISSUE TITLE
[Navi21][ROCm5.0] warpSize inconsistent between different HIP kernel building methods 

### ISSUE THREAD

**Issue by junliume** [ROCmSoftwarePlatform/MIOpen#1431]
> > > @atamazov @JehandadKhan @asroy @zjing14 @qianfengz could you take a look at CK related changes? The composable kernel-related changes in this PR might have affected accuracy:
> > > ```
> > > # ./bin/test_reduce_test --float --D 64 3 280 81 --R 0 --ReduceOp 0 --CompType 1 --N 0 --I 0 --scales 1 0
> > > ./bin/test_reduce_test --float --D 64 3 280 81 --R 0 --ReduceOp 0 --CompType 1 --N 0 --I 0 --scales 1 0 
> > > FAILED: 0.255031
> > > Iteration: 0
> > > verify_reduce_no_indices failed
> > > Input Tensor 64, 3, 280, 81
> > > Max diff: 2.27564
> > > Mismatch at 0: 1.00436 != 2.00873
> > > ```
> > 
> > 
> > Very strange! I could not reproduce the issue on MI100 and MI50 using hiprtc branch. From your test, the reduced result value get on Host is just half that of the GPU.
> 
> This might not be hiprtc related but instead ROCm 5.0 related. Removing the blocker for this PR for now.

This could be compiler issue.  I found that the `warpSize` used by my `DirectWarpSize` reduction kernels is 64, while the value of `handle.GetWavefrontWidth()` is 32.   My understand is that `warpSize` is provided as a constant by the compiler, while `handle.GetWavefrontWidth()` gets the warp size from the HIP Runtime device properties.  So the constant value maintained by the compiler under ROCM 5.0 is problematic on Navi.   The issue can be worked-around by passing the value of `handle.GetWavefrontWidth()` to the kernel and let the `DirectWarpWise` kernel to use it instead of the `warpSize`

_Originally posted by @qianfengz in https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1237#issuecomment-1041386644_

**Comment by junliume** [comment]
AFAIK Navi21 only support one warpSize. Should we enforce it for all kernels? ref to #1407 @atamazov

**Comment by qianfengz** [comment]
The issue also occurred on ROCM-4.3 ROCM-4.3.1 with my testing.  But the following codes compiled with hipcc is able to output the warpSize of 32 on Navi.  
```c++
#include <hip/hip_runtime.h>
__global__ void kernel_show_warpsize()
{
    if ( hipThreadIdx_x == 0 )
	 printf("warpSize is %d\n", warpSize); 
}; 

int main()
{

   hipLaunchKernelGGL(kernel_show_warpsize, dim3(120), dim3(256), 0, 0); 

   hipDeviceSynchronize(); 

}; 
```
Also one more important finding is that when building MIOpen with `-DMIOPEN_USE_COMGR=off`, which enables the hip kernels being compiled using the `HipBuild` path, instead of the `comgr` or `hiprtc` path, the issue does not occur. Since both `comgr` and `HipBuild` using the clang++,  it might be the compiler options used by them that causes the different values of `warpSize`.

**Comment by atamazov** [comment]
@junliume 
> AFAIK Navi21 only support one warpSize.

Navi21 supports both 64 and 32, but 32 must be the default.

> Should we enforce it for all kernels? ref to #1407 

Yes.

**Comment by atamazov** [comment]
@qianfengz 
> My understand is that warpSize is provided as a constant by the compiler,

It should be 64 for devices that do not support the `wavefrontsize64` target feature (e.g. all gfx9). For Navi21 (actually, for all known gfx10), it should be 32 by default, but can be switched to 64 (see https://llvm.org/docs/AMDGPUUsage.html#target-features).

> while handle.GetWavefrontWidth() gets the warp size from the HIP Runtime device properties.

AFAICS, HIP runtime always returns 32 for gfx10. Actual warp size depends on wavesize values set by the compiler in the code object metadata and in the kernel descriptor (these define actual wavesize for the kernel dispatch).

**Comment by qianfengz** [comment]
> @qianfengz
> 
> > My understand is that warpSize is provided as a constant by the compiler,
> 
> It should be 64 for devices that do not support the `wavefrontsize64` target feature (e.g. all gfx9). For Navi21 (actually, for all known gfx10*), it should be 32 by default, but can be switched to 64 (see https://llvm.org/docs/AMDGPUUsage.html#target-features).
> 
> > while handle.GetWavefrontWidth() gets the warp size from the HIP Runtime device properties.
> 
> AFAICS, HIP runtime always returns 32 for gfx10. Actual warp size depends on wavesize values set by the compiler in the code object metadata and in the kernel descriptor (these define actual wavesize for the kernel dispatch).

This lead to one more question,  how to let the warp size used by the host layer to be consistent with the `warpSize` seen by the kernel ?

**Comment by atamazov** [comment]
@qianfengz 
> This lead to one more question, assume the `warpSize` is correctly set, how to let the warp size used by the host layer to be consistent with the `warpSize` seen by the kernel ? 

Right now Solvers shall assume that warp size is 32 for all gfx10* GPUs and 64 for others.

[Future directions] Instead of the assumption, it may worth implementing relevant method in the handle, e.g. `GetDefaultWarpSize()`.

**Comment by junliume** [comment]
@qianfengz  The Workaround should be sufficient for ROCm 5.1. Could you implement the full "fix" as suggested above as a follow-up? Thanks!

**Comment by atamazov** [comment]
@junliume I can do this, if you wish

**Comment by junliume** [comment]
> @junliume I can do this, if you wish

Yes it would be much appreciated :)

**Comment by atamazov** [comment]
[Preliminary] The reason of the issue is a bug in PCH. Why: Bug disappears with `export MIOPEN_DEBUG_COMGR_HIP_PCH_ENFORCE=0` (COMGR on, HIPRTC off). It seems like `warpSize` is coming from PCH, but these precompiled with fixed value of 64.

We need to open Jira ticket for this issue.

I am preparing workaround that should override `warpSize` with correct value. Stay tuned.

**Comment by qianfengz** [comment]
Yes, I just tested and  verified that the bug disappears with `MIOPEN_DEBUG_COMGR_HIP_PCH_ENFORCE=0`.   By searching under /opt/rocm/hip/include,   I can find that `__hipGetPCH` is declared, but I could not find where it is defined from HIP source.  Theoretically, we should be able to check the value of `warpSize` from PCH by writing simple codes using the `comgr` API.  And by checking the implementation of `__hipGetPCH`, we should be able to know how `warpSize` is set.

**Comment by qianfengz** [comment]
A question, when adding a Unit test to test/CMakeLists.txt, how to ensure the test is executed using a specific ROCM version? I think the `inconsistent warpSize` does not occur with ROCM version between `[4.4, 5.0)`,  since `__hipGetPCH` is not available with ROCM in `[4.4, 5.0)`,  and I have verified this on ROCM `4.5.2` using `bin/test_reduce_test --all` and by checking `readelf -s -d libamdhip64.so |grep "hipGetPCH"`

**Comment by atamazov** [comment]
Additional info which may help compiler developers. Passing `-DwarpSize=32` helps for HIP builds via COMGR. This would be the best and the simplest hack. Sadly, when I build HIP kernels via HIPRTC, the following error happens:
```
MIOpen(HIP): Error [BuildHip] HIPRTC status = HIPRTC_ERROR_COMPILATION (6), source file: gridwise_generic_reduction_first_call_warpwise_reduce_partial_dims.cpp
MIOpen(HIP): Warning [BuildHip] In file included from <built-in>:1:
In file included from /long_pathname_so_that_rpms_can_package_the_debug_info/src/out/ubuntu-18.04/18.04/build/hip-on-rocclr/src/hip_rtc_gen/hipRTC_header.h:5:
In file included from /long_pathname_so_that_rpms_can_package_the_debug_info/src/external/hip-on-vdi/include/hip/hip_runtime.h:62:
In file included from /long_pathname_so_that_rpms_can_package_the_debug_info/src/external/hipamd/include/hip/amd_detail/amd_hip_runtime.h:117:
In file included from /long_pathname_so_that_rpms_can_package_the_debug_info/src/external/hipamd/include/hip/amd_detail/amd_hip_atomic.h:25:
/long_pathname_so_that_rpms_can_package_the_debug_info/src/external/hipamd/include/hip/amd_detail/amd_device_functions.h:317:22: error: expected unqualified-id
static constexpr int warpSize = 64;
                     ^
<command line>:29:18: note: expanded from here
#define warpSize 32
                 ^
1 error generated when compiling for gfx1030.
Error: Failed to compile source (from CL or HIP source to LLVM IR).
```
:arrow_up: The above denotes the source of the issue, a definition in amd_device_functions.h.

**Comment by atamazov** [comment]
@qianfengz This specific regression test is really fast so we do not need to skip it within [4.4, 5.0).

> A question, when adding a Unit test to test/CMakeLists.txt, how to ensure the test is executed using a specific ROCM version?

At build time, you can examine `MIOPEN_hip_VERSION_FLAT`/`MAJOR`/`MINOR`/`PATCH`. Example: https://github.com/ROCmSoftwarePlatform/MIOpen/blob/223b4b814676809274e105df70749fd189059a55/test/CMakeLists.txt#L1080
:warning: I do not like this idea and I recommend using it as a last resort.

**Comment by qianfengz** [comment]
I mean we need to use ROCM 4.3 or ROCM 5.0 to verify that the  issue is actually solved with our fixing.

**Comment by qianfengz** [comment]
> Additional info which may help compiler developers. Passing `-DwarpSize=32` helps for HIP builds via COMGR. This would be the best and the simplest hack. Sadly, when I build HIP kernels via HIPRTC, the following error happens:
> 
> ```
> MIOpen(HIP): Error [BuildHip] HIPRTC status = HIPRTC_ERROR_COMPILATION (6), source file: gridwise_generic_reduction_first_call_warpwise_reduce_partial_dims.cpp
> MIOpen(HIP): Warning [BuildHip] In file included from <built-in>:1:
> In file included from /long_pathname_so_that_rpms_can_package_the_debug_info/src/out/ubuntu-18.04/18.04/build/hip-on-rocclr/src/hip_rtc_gen/hipRTC_header.h:5:
> In file included from /long_pathname_so_that_rpms_can_package_the_debug_info/src/external/hip-on-vdi/include/hip/hip_runtime.h:62:
> In file included from /long_pathname_so_that_rpms_can_package_the_debug_info/src/external/hipamd/include/hip/amd_detail/amd_hip_runtime.h:117:
> In file included from /long_pathname_so_that_rpms_can_package_the_debug_info/src/external/hipamd/include/hip/amd_detail/amd_hip_atomic.h:25:
> /long_pathname_so_that_rpms_can_package_the_debug_info/src/external/hipamd/include/hip/amd_detail/amd_device_functions.h:317:22: error: expected unqualified-id
> static constexpr int warpSize = 64;
>                      ^
> <command line>:29:18: note: expanded from here
> #define warpSize 32
>                  ^
> 1 error generated when compiling for gfx1030.
> Error: Failed to compile source (from CL or HIP source to LLVM IR).
> ```
> 
> arrow_up The above denotes the source of the issue, a definition in amd_device_functions.h.

Yes,  we could not pass `-DwarpSize=64` because `warpSize` is already declared as a constant.   So it is the compiler which sets the wrong value of __AMDGCN_WAVEFRONT_SIZE ?

**Comment by qianfengz** [comment]
HIP uses the the scripts [hip_embed_pch.sh](https://github.com/ROCm-Developer-Tools/HIP/blob/rocm-4.3.x/bin/hip_embed_pch.sh) to generate the pre-compiled header. The command line to pre-compile the header file (`-E`) does not include the option `--cuda-gpu-arch=<xxx>` that is used by [HipBuildImpl](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/src/hip/hip_build_utils.cpp#L120-L124).   So `hip_embed_pch.sh` does not know which gpu arch it should generate codes for, and even though most pre-compiled header stuffs don't depend on the gpu arch, the `warpSize` does depend on gpu arch.   By adding `--cud-gpu-arch=gfx1030` and check the output file pch.cui, the scripts does generate the correct `warpSize=32` on `gfx1030`. 

So I have the following two opinions for the issue:
1.  The PCH generated by HIP is not suitable to be used by gfx1030 since HIP does not know what is the target gpu arch while it is doing PCH. May be what HIP should do is not to generating PCH, rather it should provide a API (eg. `hipGeneratePCH`) to do so. 
2. May MIOpen disable the comgr's processing of PCH completely when gfx1030 is detected ?  My understand what [PR1469](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1469) do is to override the value provided by the PCH `warpSize` symbol using the `#define warpSize 32`,  rather than disable the comgr's processing of PCH completely.

**Comment by atamazov** [comment]
@qianfengz **_Can you please open Jira ticket to ROCClr_** (as I've asked [here](https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1431#issuecomment-1072543830))? I can't do that.

And big thanks for the investigation! I believe it is worth adding this info to the Jira ticket.

> May be what HIP should do is not to generating PCH, rather it should provide a API (eg. hipGeneratePCH) to do so.

Most likely the proper solution is to generate two PCH (for warp size 32 and 64). API is not an option for us because we want to build HIP kernels when header files are NOT installed (this is the root reason of this PCH feature).

> May MIOpen disable the comgr's processing of PCH completely when gfx1030 is detected ? 

Unfortunately that would break PyTorch WHL package for gfx10.

/cc @junliume

**Comment by qianfengz** [comment]
> @qianfengz **_Can you please open Jira ticket to ROCClr_** (as I've asked [here](https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1431#issuecomment-1072543830))? I can't do that.
> 
> And big thanks for the investigation! I believe it is worth adding this info to the Jira ticket.
> 
> > May be what HIP should do is not to generating PCH, rather it should provide a API (eg. hipGeneratePCH) to do so.
> 
> Most likely the proper solution is to generate two PCH (for warp size 32 and 64). API is not an option for us because we want to build HIP kernels when header files are NOT installed (this is the root reason of this PCH feature).
> 
> > May MIOpen disable the comgr's processing of PCH completely when gfx1030 is detected ?
> 
> Unfortunately that would break PyTorch WHL package for gfx10.
> 
> /cc @junliume

@atamazov    Completely agree with you !

I have filed a JIRA   [SWDEV-329802](https://ontrack-internal.amd.com/browse/SWDEV-329802)

**Comment by atamazov** [comment]
@junliume Please lower urgency to normal or low as we have W/A merged from #1469. Thanks!

**Comment by atamazov** [comment]
The next steps: Monitor SWDEV-329802. When it is resolved in HIP mainline, use HIP version to disable W/A (change definition of `WORKAROUND_ISSUE_1431`). Then this ticket can be closed.

Further steps (dismissing the W/A when it is not necessary anymore) are out of the scope of this ticket and to be performed during routine handling of existing W/As (covered in #561).

**Comment by ppanchad-amd** [comment]
@junliume SWDEV-329802 is implemented. Can we close this ticket? Thanks!
